### PR TITLE
Performed an audit over the O3DE codebase related to chrono clocks

### LINKED ISSUE #12101: Update incorrect uses of chrono::system_clock within the codebase to use chrono::steady_clock
**Describe the bug**
Currently there 100s of places within the O3DE codebase which uses the [AZStd::chrono::system_clock](https://sourcegraph.com/search?q=context:global+repo:github%5C.com/o3de/o3de+system_clock&patternType=standard) function for measuring a duration of time, which is the incorrect clock function to use.
The AZStd/std::chrono::system_clock functions returns the wall clock time on the system which can be adjusted backwards or forwards while the system is running. This can occur if the system clock is sync'ed with an NTP server. As the system_clock is not monotonic(https://en.cppreference.com/w/cpp/chrono/system_clock) in time, using the system_clock in O3DE application is vulnerable to incorrect time steps being experienced if the system_clock changes.

Imagine there is an [AssetProcessor test](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/Tests/Asset/AssetManagerLoadingTests.cpp#L347-L354) that is running with a timeout of 5 seconds in case a test is deadlock in the test itself, but during the 5 seconds before the timeout occurs, the machine clock is synced with an NTP server which sets it backwards in time by 20 minutes(The machine was set to a really terrible clock time before). 
The timeout of the test would then become 20 minutes and 5 seconds instead of just 5 seconds.

The opposite can occur as well.
Imagine during the waiting for the 5 second timeout, 50 milliseconds the NTP server updates the system clock forward by 20 seconds.
That would cause the test to prematurely timeout where it may have been fine waiting.

What should be done is to audit the uses of std::chrono::system_clock throughout the codebase and use https://en.cppreference.com/w/cpp/chrono/steady_clock.

The `std::chrono::steady_clock` class is monotonic(It only counts upwards and is never adjusted once the system is running) and therefore not subject to wall clock sync issues.

**Steps to reproduce**
1. Add the following test to the [ChronoTests.cpp ](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/Tests/AZStd/ChronoTests.cpp)file
```c++

AZStd::chrono::time_point startTime = AZStd::chrono::system_clock::now();
AZStd::chrono::seconds timeout(60);

while (AZStd::chrono::system_clock::now() - startTime <= timeout)
{
    // Do something to make sure this block isn't optimized out
}
```

2. Open the machines Date and Time settings(this is dependent on the OS being run)
3. Running only the added unit test.  The `--gtest_filter=*<TestName>` option can be used for this
4. During the 60 seconds that the test is timed box for, adjust the system clock backwards an hour
5. Wait until the original 60 seconds for the timeout of the test is 

**Expected behavior**
Test ends after 60 seconds

**Actual behavior**
Test runs for an hour and 60 seconds

**Found in Branch**
development

**Desktop/Device (please complete the following information):**
 - OS: All

**Additional context**
See this StackOverflow post by Howard Hinnant as to why steady_clock should be used for timing short durations
https://stackoverflow.com/a/64084510

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Locations that were using `AZStd::chrono::system_clock` to time durations were switched out to use AZStd::chrono::steady_clock, which is monotonic.

As the system_clock isn't monotonic it isn't suitable for measuring time before and after an event as it can be adjusted forwards and backwards See microsoft API for [chrono::system_clock](https://learn.microsoft.com/en-us/cpp/standard-library/system-clock-structure?view=msvc-170)

Also replaced use of high_resolution_clock with steady_clock, since high_resolution is an alias to either system_clock or steady_clock and all uses of it was for measuring a duration of time.

fixes #12101

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

## How was this PR tested?

Ran the registered CTest
Loaded up a Sample level in the Editor and ran through it.

### CODE REVIEW

**Review by cgalvan — APPROVED** [review]
Nice change!

**Review by rgba16f — APPROVED** [review]

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/Process/Scheduler/TestImpactProcessScheduler.h [hunk 1]
```diff
@@ -60,7 +60,7 @@ namespace TestImpact
         AZStd::function<ProcessCallbackResult(
             ProcessId processId,
             LaunchResult launchResult,
-            AZStd::chrono::high_resolution_clock::time_point createTime)>;
+            AZStd::chrono::steady_clockeTime)>;
```
Comment (amzn-phist) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/Process/Scheduler/TestImpactProcessScheduler.h:63]: Fix this: `AZStd::chrono::steady_clock::time_point createTime`

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Process/Scheduler/TestImpactProcessScheduler.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ namespace TestImpact
         AZStd::optional<ProcessStdContentCallback> m_processStdContentCallback;
         AZStd::optional<AZStd::chrono::milliseconds> m_processTimeout;
         AZStd::optional<AZStd::chrono::milliseconds> m_scheduleTimeout;
-        AZStd::chrono::high_resolution_clock::time_point m_startTime;
+        AZStd::chrono::steady_clockint m_startTime;
```
Comment (amzn-phist) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Process/Scheduler/TestImpactProcessScheduler.cpp:50]: Fix this: `AZStd::chrono::steady_clock::time_point`

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/Process/Scheduler/TestImpactProcessScheduler.h (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/Process/Scheduler/TestImpactProcessScheduler.h:63]: What happened here?

**Review by amzn-phist — APPROVED** [review]
Thanks

**Review by AMZN-puvvadar — APPROVED** [review]
Looks like just some unit test updates in MPCompression for Networking which seem fine. I'd get either @allisaurus or @lmbr-pip to weigh in on the AWS Gem changes.

**Comment by kberg-amzn** [comment]
Less a review on the changes themselves, but more a question for you.

Given how undisciplined we appear to have been in our timer usage, how far off are we from realizing t_scale or SetGlobalTimeDilation() like functionality, similar to what exists in unreal or cry or any other major engine? Time scaling functionality was built into AZ::ITime intentionally since it's a critically important feature for some of our simulation customers, but seeing all the AZ/std::chrono usage here is really worrying for me.

**Comment by lumberyard-employee-dm** [comment]
> Less a review on the changes themselves, but more a question for you.
> 
> Given how undisciplined we appear to have been in our timer usage, how far off are we from realizing t_scale or SetGlobalTimeDilation() like functionality, similar to what exists in unreal or cry or any other major engine? Time scaling functionality was built into AZ::ITime intentionally since it's a critically important feature for some of our simulation customers, but seeing all the AZ/std::chrono usage here is really worrying for me.

I think a SetGlobalTimeDilation function can be made, if we update the appropriate touch points. 
Such as the [ScriptTimePoint](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h#L30-L33), the [Game Tick system](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Component/TickBus.h#L153-L172)) and [ITime](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Time/ITime.h#L138-L178).

Despite all of the incorrect usages of system_clock over steady_clock in the codebase, all of it except for the ScriptTimePoint.h/.cpp and ITime.h changes were related to real time application operations not affected by the game time scale. Most of it is easier used to time a particular sequence of C++ function calls(miscellaneous use all over the code base), set a timeout in real time(AssetProcessor Connection, etc...), or set a deadline for an operation to occur(Streamer, condition_variable, semaphore)

For global time dilation we can hide that behind the ScriptTimePoint for scripting.
For C++ code the TimeSystem class has the [GetLastSimulationTickTime](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Time/TimeSystem.cpp#L136-L166) function which already accounts for the simulation time scale scalar.

So for time dilation, as long as the ITime interface GetSimulation* functions is used to query time, we are good on the C++ front.

For scripting though... I would recommend updating the [ScriptTimePoint](https://github.com/o3de/o3de/blob/f0ba861fd87a56de93f1830348dff4723b1f0a14/Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h#L30-L31) class to use the ITime interface for querying the simulation time(instead of defaulting to real time based on std::chrono).

**Comment by kberg-amzn** [comment]
> That allows use to fix up all the locations which are passing in real time as it would be a compiler error. It also encapsulates querying of the time to use the ITime API

Excellent. We were just talking about the importance of time scaling in our robotics program review this morning in fact, so it's good to hear most of our current system clock touchpoints don't impact simulation time scale. We should get your suggestion for the ScriptTimePoint constructor in front of @lsemp3d so it can get looked at promptly.

**Comment by lsemp3d** [comment]
> > That allows use to fix up all the locations which are passing in real time as it would be a compiler error. It also encapsulates querying of the time to use the ITime API
> 
> Excellent. We were just talking about the importance of time scaling in our robotics program review this morning in fact, so it's good to hear most of our current system clock touchpoints don't impact simulation time scale. We should get your suggestion for the ScriptTimePoint constructor in front of @lsemp3d so it can get looked at promptly.

I have no objections, ScriptTimePoint is legacy code that predates LY and in all these years I personally have never found a use for it, deltaTime being sufficient for most time operations we do.

**Review by lsemp3d — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
> > > That allows use to fix up all the locations which are passing in real time as it would be a compiler error. It also encapsulates querying of the time to use the ITime API
> > 
> > 
> > Excellent. We were just talking about the importance of time scaling in our robotics program review this morning in fact, so it's good to hear most of our current system clock touchpoints don't impact simulation time scale. We should get your suggestion for the ScriptTimePoint constructor in front of @lsemp3d so it can get looked at promptly.
> 
> I have no objections, ScriptTimePoint is legacy code that predates LY and in all these years I personally have never found a use for it, deltaTime being sufficient for most time operations we do. I'd ask the question, is there something that relies on it? can we red-code it?

The [ComponentApplication::OnTick](https://github.com/o3de/o3de/blob/f0ba861fd87a56de93f1830348dff4723b1f0a14/Code/Framework/AzCore/AzCore/Component/ComponentApplication.cpp#L1375-L1392) function uses the TimeSystem for providing the delta time and ScriptTimePoint.
We should probably get rid of it then if we don't have a use for it.
Though for the OnTick function, it accepts a raw `float` which allows multiple systems to incorrectly supply real time.
Perhaps the TickEvents::OnEvents system should only accept a "GameTimePoint" type forces the callers to supply a time scale a particular unit of time(it could be microseconds).
For example
```
struct GameTimePoint
{
    GameTimePoint(AZStd::chrono::microseconds deltaTime, float scaleOfDeltaTime);
};

**Review by lmbr-pip — APPROVED** [review]
AWS changes look good (responding to @AMZN-puvvadar) request.

Asking Asset Pipeline experts just to look over AP changes, but they look ok to me.

**Comment by lsemp3d** [comment]
> > > > That allows use to fix up all the locations which are passing in real time as it would be a compiler error. It also encapsulates querying of the time to use the ITime API
> > > 
> > > 
> > > Excellent. We were just talking about the importance of time scaling in our robotics program review this morning in fact, so it's good to hear most of our current system clock touchpoints don't impact simulation time scale. We should get your suggestion for the ScriptTimePoint constructor in front of @lsemp3d so it can get looked at promptly.
> > 
> > 
> > I have no objections, ScriptTimePoint is legacy code that predates LY and in all these years I personally have never found a use for it, deltaTime being sufficient for most time operations we do. I'd ask the question, is there something that relies on it? can we red-code it?
> 
> The [ComponentApplication::OnTick](https://github.com/o3de/o3de/blob/f0ba861fd87a56de93f1830348dff4723b1f0a14/Code/Framework/AzCore/AzCore/Component/ComponentApplication.cpp#L1375-L1392) function uses the TimeSystem for providing the delta time and ScriptTimePoint. We should probably get rid of it then if we don't have a use for it. Though for the OnTick function, it accepts a raw `float` which allows multiple systems to incorrectly supply real time. Perhaps the TickEvents::OnEvents system should only accept a "GameTimePoint" type forces the callers to supply a time scale a particular unit of time(it could be microseconds). For example
> 
> ```
> struct GameTimePoint
> {
>     GameTimePoint(AZStd::chrono::microseconds deltaTime, float scaleOfDeltaTime);
> };
> ```

I prefer a raw float because it allows users to scale script operations by the frame duration resulting in smooth interpolations and time advancement, being a float it makes the script operations really trivial, no need to reflect, alias or create variables of a more advanced type, this just makes things less intuitive. From a scripting perspective, `float deltaTime` is not broken, not sure we need to change it. Passing in the time dilation scale, OR adding a method to query the time dliation is definitely useful though, not all scripts need to respect time dilation, but giving developers that options is good.

**Review by amzn-mike — APPROVED** [review]

**Comment by kberg-amzn** [comment]
> I prefer a raw float because it allows users to scale script operations by the frame duration resulting in smooth interpolations and time advancement, being a float it makes the script operations really trivial, no need to reflect, alias or create variables of a more advanced type, that just makes things less intuitive. From a scripting perspective, `float deltaTime` is not broken, not sure we need to change it. Passing in the time dilation scale, OR adding a method to query the time dliation is definitely useful though, not all scripts need to respect time dilation, but giving developers that options is good.

We don't want developers to have to remember to sample a time dilation value and apply it every time they create a script that should respect simulation time; that seems a recipe for disaster. Usually in other engines high-level simulation systems _only_ know about simulation time, system time isn't accessible. This just prevents coding and content errors from accumulating that destroy the ability of a project to alter simulation speed.

File: Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h [hunk 3]
```diff
@@ -17,26 +17,26 @@ namespace AZ
 {
     class ReflectContext;
 
-    AZ_TYPE_INFO_SPECIALIZE(AZStd::chrono::system_clock::time_point, "{5C48FD59-7267-405D-9C06-1EA31379FE82}");
+    AZ_TYPE_INFO_SPECIALIZE(AZStd::chrono::steady_clock::time_point, "{5C48FD59-7267-405D-9C06-1EA31379FE82}");
 
 
-    //! Wrapper that reflects a AZStd::chrono::system_clock::time_point to script.
+    //! Wrapper that reflects a AZStd::chrono::steady_clock::time_point to script.
```
Comment (AMZN-stankowi) [inline Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h:23]: Are you sure you want to change ScriptTimePoint to use steady_clock instead of system_clock?

Steady_clock's documentation https://en.cppreference.com/w/cpp/chrono/steady_clock says it's not meant to be used to represent a wall clock time, it could be time since boot. This says to mean it wouldn't be useful for tracking times across machines or boot sessions, as well as potentially not useful for printing/displaying the current time.

I'm ont sure how ScriptTimePoint is actually used, but if there are use cases where time is either cached between rebooting a machine, or meant to show a clock's time, it seems better to use system_clock.

**Comment by lumberyard-employee-dm** [comment]
> > I prefer a raw float because it allows users to scale script operations by the frame duration resulting in smooth interpolations and time advancement, being a float it makes the script operations really trivial, no need to reflect, alias or create variables of a more advanced type, that just makes things less intuitive. From a scripting perspective, `float deltaTime` is not broken, not sure we need to change it. Passing in the time dilation scale, OR adding a method to query the time dliation is definitely useful though, not all scripts need to respect time dilation, but giving developers that options is good.
> 
> We don't want developers to have to remember to sample a time dilation value and apply it every time they create a script that should respect simulation time; that seems a recipe for disaster. Usually in other engines high-level simulation systems _only_ know about simulation time, system time isn't accessible. This just prevents coding and content errors from accumulating that destroy the ability of a project to alter simulation speed.

The time diliation value would be baked into the deltaTime. The other parameter is to provide the callee with the information as to what the time diliation value was when the call was made.

File: Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.h [hunk 4]
```diff
@@ -769,7 +769,7 @@ namespace ScriptCanvasEditor
         void ClearStaleSaves();
         bool IsRecentSave(const SourceHandle& handle) const;
         void MarkRecentSave(const SourceHandle& handle);
-        AZStd::recursive_mutex m_mutex; 
-        AZStd::unordered_map <AZStd::string, AZStd::chrono::system_clock::time_point> m_saves;
+        AZStd::recursive_mutex m_mutex;
+        AZStd::unordered_map <AZStd::string, AZStd::chrono::steady_clock::time_point> m_saves;
```
Comment (AMZN-stankowi) [inline Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.h:773]: This isn't cached and re-used across sessions of the tool, is it?

If so, steady_clock doesn't seem useful for storing times between rebooting a machine. If not, then it's fine.

**Review by AMZN-stankowi — APPROVED** [review]
I'm pretty sure you've verified that the couple places I called out in this review are used safely with steady_clock, but I'll post the review feedback anyways in case there's a gotcha there.

The only real concern I would have with this is it might impact testing that was relying on clock tricks to force edge cases or long timeouts. I think it's fine putting this in as-is, and if those come up we can make changes to allow system clock changing (or other solutions) for specific cases we want to test timeouts.

File: Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h (hunk 3 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Script/ScriptTimePoint.h:23]: The ScriptTimePoint is supplied via the OnTick method. It is meant to count clock ticks since the previous call to OnTick

File: Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.h (hunk 4 above)
Comment (lumberyard-employee-dm) [reply Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.h:773]: This is used to update the file tracker state of an open script canvas file that has been [saved in the previous 2 seconds](https://github.com/o3de/o3de/blob/development/Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.cpp#L4438-L4447).
The file tracker state is used to indicate whether a file has been [modified or saved](https://github.com/o3de/o3de/blob/development/Gems/ScriptCanvas/Code/Editor/View/Windows/MainWindow.cpp#L1049-L1060) and update the taskbar with that information. Such foo.scriptcanvas*, when it has been modified, but not saved yet.

**Comment by lumberyard-employee-dm** [comment]
> I'm pretty sure you've verified that the couple places I called out in this review are used safely with steady_clock, but I'll post the review feedback anyways in case there's a gotcha there.
> 
> The only real concern I would have with this is it might impact testing that was relying on clock tricks to force edge cases or long timeouts. I think it's fine putting this in as-is, and if those come up we can make changes to allow system clock changing (or other solutions) for specific cases we want to test timeouts.

Any kind of clock tricks in testing(which we currently don't have) to force a timeout to occur sooner, would involve changing the time on the machine, which iisn't proper testing methodology.
Test should never change the actual machine time.
